test/bench_sub_async.c

```diff
@@ -0,0 +1,500 @@
+    for (int i = 0; i < env->numSubs; i++)
+        env->subs[i].env = env; // set the environment to access it in the callbacks.
+
+    int pid = _startServer("nats://127.0.0.1:4222", NULL, true);
+    if (pid == NATS_INVALID_PID)
+        return NATS_ERR;
+
+    natsStatus s = nats_Open(-1);
+    IFOK(s, natsNUID_Next(subject, NUID_BUFFER_LEN + 1));
+    IFOK(s, natsOptions_Create(&opts));
+    IFOK(s, nats_SetMessageDeliveryPoolSize(env->threads.max));
+    IFOK(s, natsOptions_SetErrorHandler(opts, _onError, NULL));
+    IFOK(s, natsOptions_UseGlobalMessageDelivery(opts, env->threads.useGlobalDelivery));
+
+    IFOK(s, natsConnection_Connect(&nc, opts));
+
+    for (int i = 0; i < env->numSubs; i++)
+    {
+        IFOK(s, natsConnection_Subscribe(&(env->subs[i].sub), nc, subject, _onMessage, &env->subs[i]));
+        IFOK(s, natsSubscription_SetPendingLimits(env->subs[i].sub, -1, -1));
+        IFOK(s, natsSubscription_AutoUnsubscribe(env->subs[i].sub, env->numPubMessages));
+        IFOK(s, natsSubscription_SetOnCompleteCB(env->subs[i].sub, _onComplete, &env->subs[i]));
+    }
+
+    start = nats_Now();
+
+    // Publish or inject the messages!
+    IFOK(s, env->pubf(nc, subject, env));
+
+    while (s == NATS_OK)
+    {
+        bool done = true;
+        for (int i = 0; i < env->numSubs; i++)
+        {
+            // threads don't touch this, should be safe
+            if (natsSubscription_IsValid(env->subs[i].sub))
+            {
+                done = false;
+                break;
+            }
+        }
+
+        nats_Sleep(10);
+        if (done)
+            break;
+    }
+
+    b = w = a = 0;
+    if (s == NATS_OK)
+    {
+        for (int i = 0; i < env->numSubs; i++)
+        {
+            if (env->subs[i].sum != expectedSum)
+            {
+                s = NATS_ERR;
+                fprintf(stderr, "Error: sum is %" PRId64 " for sub %d, expected %" PRId64 "\n", env->subs[i].sum, i, expectedSum);
+                break;
+            }
+            if (env->subs[i].xor != expectedXOR)
+            {
+                fprintf(stderr, "Error: xor is %" PRId64 " for sub %d, expected %" PRId64 "\n", env->subs[i].xor, i, expectedXOR);
+                s = NATS_ERR;
+                break;
+            }
+            if ((int)(env->subs[i].count) != env->numPubMessages)
+            {
+                fprintf(stderr, "Error: count is %" PRId64 " for sub %d, expected %d\n", env->subs[i].count, i, env->numPubMessages);
+                s = NATS_ERR;
+                break;
+            }
+
+            int64_t dur = env->subs[i].closedTimestamp - start;
+            if (dur > w)
+                w = dur;
+            if ((dur < b) || (b == 0))
+                b = dur;
+            a += dur;
+        }
+    }
+
+    // cleanup
+    for (int i = 0; i < env->numSubs; i++)
+        natsSubscription_Destroy(env->subs[i].sub);
+    natsConnection_Destroy(nc);
+    natsOptions_Destroy(opts);
+    nats_CloseAndWait(0);
```

Comment:
You should move this after `_stopServer` since we use mutex and hashtable there.